// dependency-lint/dist/formatters/json_formatter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _error_messages = _interopRequireDefault(require("./error_messages"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class JsonFormatter {
  // stream - writable stream to send output
  constructor({
    stream: stream
  }) {
    this.stream = stream;
  } // Prints the result to its stream


  print({
    fixes: fixes,
    results: results
  }) {
    if (!fixes) {
      fixes = {};
    }

    const data = _lodash.default.mapValues(results, (modules, type) => _lodash.default.map(modules, function (module) {
      const fixed = _lodash.default.includes(fixes[type], module.name);

      const error = _error_messages.default[module.error];
      return _lodash.default.assign({}, module, {
        error: error,
        fixed: fixed
      });
    }));

    return this.stream.write(JSON.stringify(data, null, 2), 'utf8');
  }

}

exports.default = JsonFormatter;
