// dependency-lint/dist/linter/used_module_finder/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _executed_module_finder = _interopRequireDefault(require("./executed_module_finder"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _required_module_finder = _interopRequireDefault(require("./required_module_finder"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class UsedModuleFinder {
  constructor(config) {
    this.executedModuleFinder = new _executed_module_finder.default(config.executedModules);
    this.requiredModuleFinder = new _required_module_finder.default(config.requiredModules);
  }

  async find({
    dir: dir,
    packageJson: packageJson
  }) {
    return this.normalizeModules((await _bluebird.default.all([this.executedModuleFinder.find({
      dir: dir,
      packageJson: packageJson
    }), this.requiredModuleFinder.find(dir)])));
  }

  normalizeModules(...modules) {
    const result = {};

    for (const {
      name: name,
      file: file,
      script: script
    } of _lodash.default.flattenDeep(modules)) {
      if (!result[name]) {
        result[name] = {
          name: name,
          files: [],
          scripts: []
        };
      }

      if (file) {
        result[name].files.push(file);
      }

      if (script) {
        result[name].scripts.push(script);
      }
    }

    return _lodash.default.values(result);
  }

}

exports.default = UsedModuleFinder;
