// dependency-lint/dist/linter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _dependency_linter = _interopRequireDefault(require("./dependency_linter"));

var _installed_module_validator = _interopRequireDefault(require("./installed_module_validator"));

var _used_module_finder = _interopRequireDefault(require("./used_module_finder"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Linter {
  constructor(config) {
    this.dependencyLinter = new _dependency_linter.default(config);
    this.installedModuleValidater = new _installed_module_validator.default();
    this.usedModuleFinder = new _used_module_finder.default(config);
  }

  getListedModules(packageJson) {
    const result = {};
    ['dependencies', 'devDependencies'].forEach(value => result[value] = _lodash.default.keys(packageJson[value]));
    return result;
  }

  async lint({
    dir: dir,
    packageJson: packageJson
  }) {
    await this.installedModuleValidater.validate({
      dir: dir,
      packageJson: packageJson
    });
    const usedModules = await this.usedModuleFinder.find({
      dir: dir,
      packageJson: packageJson
    });
    const listedModules = this.getListedModules(packageJson);
    return this.dependencyLinter.lint({
      listedModules: listedModules,
      usedModules: usedModules
    });
  }

}

exports.default = Linter;
